File: libs/vulnfounder-core/parsers/rust/unit_generator.py
"""
Stage 4: Unit Generator for Rust

Creates self-contained analysis units with dependency context.
"""

from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List, Optional, Set

from utilities.file_io import write_json
from core.file_boundary import neutralize_boundaries


class UnitGenerator:
    """Generates analysis units from call graph data."""

    # File boundary marker using Rust comment syntax.
    FILE_BOUNDARY = "\n\n// ========== File Boundary ==========\n\n"

    def __init__(
        self,
        call_graph_output: Dict[str, Any],
        repo_path: str,
        dependency_depth: int = 3,
    ):
        self.functions = call_graph_output.get("functions", {})
        self.classes = call_graph_output.get("classes", {})
        self.call_graph = call_graph_output.get("call_graph", {})
        self.reverse_call_graph = call_graph_output.get("reverse_call_graph", {})
        self.repository = repo_path
        self.dependency_depth = dependency_depth

    def generate(self, name: Optional[str] = None) -> tuple[Dict[str, Any], Dict[str, Any]]:
        """
        Generate analysis units.

        Returns:
            (dataset.json, analyzer_output.json)
        """
        units = []
        dataset_name = name or Path(self.repository).name

        for func_id, func_info in self.functions.items():
            # Seed-only: a synthesized `fuzz_target!` harness seeds reachability
            # (its decode callees are kept and analyzed), but the harness itself
            # is instrumentation, not an analysis target — its lifted body is
            # synthetic (references the dropped closure param) and would ship
            # non-compiling code to Stage-1 and re-analyze the decoder. It stays
            # in analyzer_output below so the call graph remains symmetric.
            if func_info.get("synthetic_harness"):
                continue
            unit = self._generate_unit(func_id, func_info)
            units.append(unit)

        by_type: Dict[str, int] = {}
        units_with_upstream = 0
        units_with_downstream = 0
        total_upstream = 0
        total_downstream = 0

        for unit in units:
            unit_type = unit["unit_type"]
            by_type[unit_type] = by_type.get(unit_type, 0) + 1

            dep_meta = unit["code"]["dependency_metadata"]
            if dep_meta["total_upstream"] > 0:
                units_with_upstream += 1
                total_upstream += dep_meta["total_upstream"]
            if dep_meta["total_downstream"] > 0:
                units_with_downstream += 1
                total_downstream += dep_meta["total_downstream"]

        avg_upstream = total_upstream / len(units) if units else 0
        avg_downstream = total_downstream / len(units) if units else 0

        dataset = {
            "name": dataset_name,
            "repository": self.repository,
            "units": units,
            "statistics": {
                "total_units": len(units),
                "by_type": by_type,
                "units_with_upstream": units_with_upstream,
                "units_with_downstream": units_with_downstream,
                "units_enhanced": len(
                    [u for u in units if u["code"]["primary_origin"]["deps_inlined"]]
                ),
                "avg_upstream": round(avg_upstream, 2),
                "avg_downstream": round(avg_downstream, 2),
            },
            "metadata": {
                "generator": "rust_unit_generator.py",
                "generated_at": datetime.now().isoformat(),
                "dependency_depth": self.dependency_depth,
            },
        }

        analyzer_output = {
            "repository": self.repository,
            "functions": {
                func_id: {
                    "name": func_info["name"],
                    "unitType": func_info["unit_type"],
                    "code": func_info["code"],
                    "filePath": func_info["file_path"],
                    "startLine": func_info["start_line"],
                    "endLine": func_info["end_line"],
                    "isExported": func_info.get("is_exported", False),
                    "parameters": func_info.get("parameters", []),
                    "className": func_info.get("class_name"),
                }
                for func_id, func_info in self.functions.items()
            },
            "call_graph": self.call_graph,
            "reverse_call_graph": self.reverse_call_graph,
        }

        return dataset, analyzer_output

    def _generate_unit(self, func_id: str, func_info: Dict[str, Any]) -> Dict[str, Any]:
        upstream = self._get_dependencies(func_id, self.call_graph, self.dependency_depth)
        downstream = self._get_dependencies(func_id, self.reverse_call_graph, self.dependency_depth)

        direct_calls = self.call_graph.get(func_id, [])
        direct_callers = self.reverse_call_graph.get(func_id, [])

        primary_code, files_included = self._build_enhanced_code(func_id, func_info, upstream)

        original_length = len(func_info.get("code", ""))
        enhanced_length = len(primary_code)

        return {
            "id": func_id,
            "unit_type": func_info["unit_type"],
            "code": {
                "primary_code": primary_code,
                "primary_origin": {
                    "file_path": func_info["file_path"],
                    "start_line": func_info["start_line"],
                    "end_line": func_info["end_line"],
                    "function_name": func_info["name"],
                    "class_name": func_info.get("class_name"),
                    "deps_inlined": len(upstream) > 0,
                    "files_included": files_included,
                    "original_length": original_length,
                    "enhanced_length": enhanced_length,
                },
                "dependencies": [],
                "dependency_metadata": {
                    "depth": self.dependency_depth,
                    "total_upstream": len(upstream),
                    "total_downstream": len(downstream),
                    "direct_calls": len(direct_calls),
                    "direct_callers": len(direct_callers),
                },
            },
            "ground_truth": {
                "status": "UNKNOWN",
                "vulnerability_types": [],
                "issues": [],
                "annotation_source": None,
                "annotation_key": None,
                "notes": None,
            },
            "metadata": {
                "parameters": func_info.get("parameters", []),
                "generator": "rust_unit_generator.py",
                "direct_calls": direct_calls,
                "direct_callers": direct_callers,
                # Parity with the Swift parser's unit metadata: carry the function's
                # attributes/decorators onto dataset units.
                "decorators": func_info.get("decorators", []),
            },
        }

    def _get_dependencies(
        self, func_id: str, graph: Dict[str, List[str]], max_depth: int
    ) -> Set[str]:
        dependencies: Set[str] = set()
        current_level = {func_id}

        for _ in range(max_depth):
            next_level: Set[str] = set()
            for fid in current_level:
                for dep in graph.get(fid, []):
                    if dep not in dependencies and dep != func_id:
                        dependencies.add(dep)
                        next_level.add(dep)
            current_level = next_level
            if not current_level:
                break

        return dependencies

    def _build_enhanced_code(
        self, func_id: str, func_info: Dict[str, Any], upstream: Set[str]
    ) -> tuple[str, List[str]]:
        primary_code = func_info.get("code", "")
        files_included = [func_info["file_path"]]

        if not upstream:
            return primary_code, files_included

        deps_by_file: Dict[str, List[str]] = {}
        for dep_id in upstream:
            dep_info = self.functions.get(dep_id)
            if dep_info:
                file_path = dep_info["file_path"]
                deps_by_file.setdefault(file_path, []).append(dep_id)

        # Every string appended below is source from the scanned repository,
        # so each is neutralized before entering the blob (see
        # core/file_boundary.neutralize_boundaries): a file containing a
        # boundary-shaped line would otherwise split the unit and hide
        # everything after it from analysis.
        code_parts = [neutralize_boundaries(primary_code)]

        for file_path, dep_ids in deps_by_file.items():
            if file_path == func_info["file_path"]:
                for dep_id in dep_ids:
                    dep_info = self.functions.get(dep_id)
                    if dep_info:
                        code_parts.append(neutralize_boundaries(dep_info.get("code", "")))
            else:
                if file_path not in files_included:
                    files_included.append(file_path)
                file_code = []
                for dep_id in dep_ids:
                    dep_info = self.functions.get(dep_id)
                    if dep_info:
                        file_code.append(neutralize_boundaries(dep_info.get("code", "")))
                if file_code:
                    code_parts.append(self.FILE_BOUNDARY + "\n".join(file_code))

        return "\n\n".join(code_parts), files_included

    def save_results(
        self,
        output_dir: str,
        dataset: Dict[str, Any],
        analyzer_output: Dict[str, Any],
    ) -> None:
        output_path = Path(output_dir)
        output_path.mkdir(parents=True, exist_ok=True)

        write_json(output_path / "dataset.json", dataset)
        write_json(output_path / "analyzer_output.json", analyzer_output)

File: libs/vulnfounder-core/parsers/javascript/unit_generator.js
#!/usr/bin/env node
/**
 * Unit Generator
 *
 * Creates analysis units for ALL functions extracted from a repository.
 * Supports both static and LLM-enhanced dependency resolution.
 *
 * Usage:
 *   node unit_generator.js <typescript_analyzer_output.json> [--output <file>] [--depth <N>] [--llm]
 *
 * Input:
 *   - typescript_analyzer_output.json: Output from typescript_analyzer.js (batch mode)
 *
 * Output (JSON):
 *   {
 *     "name": "dataset_name",
 *     "repository": "/path/to/repo",
 *     "units": [
 *       {
 *         "id": "file.ts:functionName",
 *         "unit_type": "route_handler" | "middleware" | "model" | "utility" | "class_method" | "function",
 *         "code": {
 *           "primary_code": "...",
 *           "primary_origin": {...},
 *           "upstream_dependencies": [{ "id": "...", "code": "...", ... }],  // Functions this calls
 *           "downstream_callers": [{ "id": "...", "code": "...", ... }],     // Functions that call this
 *           "dependency_metadata": { "depth": 3, "total_upstream": 5, "total_downstream": 2 }
 *         },
 *         "data_flow": {                    // Optional: from LLM analysis
 *           "inputs": [...],
 *           "outputs": [...],
 *           "tainted_variables": [...],
 *           "security_relevant_flows": [...]
 *         },
 *         "route": null,  // Populated for route_handlers if route info available
 *         "ground_truth": { "status": "UNKNOWN" },
 *         "metadata": {...}
 *       }
 *     ],
 *     "statistics": {
 *       "total_units": 150,
 *       "by_type": { "route_handler": 20, "function": 100, ... },
 *       "call_graph": { "total_edges": 500, "avg_out_degree": 2.5, ... }
 *     }
 *   }
 */

const fs = require('fs');
const path = require('path');
const { DependencyResolver } = require('./dependency_resolver');

// File boundary marker for enhanced code (module-level for parity with the
// python/php/c/ruby parsers, so external consumers can import the canonical
// marker instead of re-defining it and risking silent drift).
const FILE_BOUNDARY = '\n\n// ========== File Boundary ==========\n\n';

// A whole boundary line, either comment style, anchored to line starts. Mirrors
// core/file_boundary.py's _BOUNDARY_LINE — the two must stay in step, which
// tests/test_hostile_repo.py enforces by requiring every producer to neutralize.
const BOUNDARY_LINE_RE = /^[ \t]*(?:#|\/\/)?[ \t]*={10} File Boundary ={10}[ \t]*$/gm;

/**
 * Defang boundary-shaped lines in untrusted source before concatenation.
 *
 * Scanned source can contain a line that mimics the separator. Once files are
 * joined, a forged marker is byte-identical to ours, so the split cuts the unit
 * there and everything after it is relabelled "do NOT analyze" in the prompt —
 * one comment line hides a vulnerability from both analysis stages. The consumer
 * cannot distinguish them after the fact, so this has to happen here.
 *
 * @param {string} source Untrusted source from the scanned repository.
 * @returns {string} Source with boundary-shaped lines replaced.
 */
function neutralizeBoundaries(source) {
    if (!source) return source;
    return source.replace(
        BOUNDARY_LINE_RE,
        '// [openant] boundary-shaped line from scanned source, neutralized'
    );
}

class UnitGenerator {
    constructor(repoPath, datasetName = null, options = {}) {
        this.repoPath = repoPath;
        this.datasetName = datasetName || path.basename(repoPath);
        this.maxDepth = options.maxDepth || 3;
        this.units = [];
        this.resolver = null;  // Initialized in generateUnits
        this.statistics = {
            totalUnits: 0,
            byType: Object.create(null),
            callGraph: {}
        };
    }

    /**
     * Generate analysis units from TypeScript analyzer output
     * @param {Object} analyzerOutput - Output from TypeScriptAnalyzer.analyzeFiles()
     * @param {Object} routeInfo - Optional route information to tag route handlers
     */
    generateUnits(analyzerOutput, routeInfo = null) {
        const { functions } = analyzerOutput;

        // Initialize DependencyResolver and build call graph
        this.resolver = new DependencyResolver(analyzerOutput, { maxDepth: this.maxDepth });
        this.resolver.buildCallGraph();

        // Store call graph statistics
        const resolverStats = this.resolver.getStatistics();
        this.statistics.callGraph = {
            totalEdges: resolverStats.totalEdges,
            avgOutDegree: resolverStats.avgOutDegree,
            avgInDegree: resolverStats.avgInDegree,
            maxOutDegree: resolverStats.maxOutDegree,
            maxInDegree: resolverStats.maxInDegree,
            isolatedFunctions: resolverStats.isolatedFunctions
        };

        // Track units with dependencies
        this.statistics.unitsWithUpstream = 0;
        this.statistics.unitsWithDownstream = 0;

        // Build a map of route handlers if route info is provided
        const routeHandlerMap = this._buildRouteHandlerMap(routeInfo);

        // Process all functions
        const functionIds = Object.keys(functions);
        for (const functionId of functionIds) {
            const funcData = functions[functionId];
            const unit = this._createUnit(functionId, funcData, routeHandlerMap);
            this.units.push(unit);
            this._updateStatistics(unit);
        }

        return {
            name: this.datasetName,
            repository: this.repoPath,
            units: this.units,
            statistics: this.statistics,
            metadata: {
                generator: 'unit_generator.js',
                generated_at: new Date().toISOString(),
                dependency_depth: this.maxDepth
            }
        };
    }

    /**
     * Update statistics for a unit
     */
    _updateStatistics(unit) {
        this.statistics.totalUnits++;
        const unitType = unit.unit_type;
        this.statistics.byType[unitType] = (this.statistics.byType[unitType] || 0) + 1;

        const depMeta = unit.code.dependency_metadata || {};
        if (depMeta.total_upstream > 0) {
            this.statistics.unitsWithUpstream++;
        }
        if (depMeta.total_downstream > 0) {
            this.statistics.unitsWithDownstream++;
        }
    }

    /**
     * Build a map of route handlers from route information
     */
    _buildRouteHandlerMap(routeInfo) {
        const map = new Map();
        if (!routeInfo || !routeInfo.routes) return map;

        for (const route of routeInfo.routes) {
            const handler = route.handler;
            const file = route.file;

            // Create keys for matching
            // Pattern: "file.ts:handlerName" or "file.ts:ClassName.methodName"
            if (handler && file) {
                const key = `${file}:${handler}`;
                map.set(key, route);

                // Also store just by handler name for simpler matching
                map.set(handler, route);
            }
        }

        return map;
    }

    /**
     * Assemble enhanced code with all dependencies using file boundary markers
     * Matches DVNA enhanced dataset format expected by experiment.py
     */
    _assembleEnhancedCode(funcData, upstreamDependencies, downstreamCallers) {
        const parts = [];
        const includedCode = new Set();

        // Add primary code first
        parts.push(funcData.code);
        includedCode.add(funcData.code);

        // Add upstream dependencies (functions this calls)
        for (const dep of upstreamDependencies) {
            if (dep.code && !includedCode.has(dep.code)) {
                parts.push(dep.code);
                includedCode.add(dep.code);
            }
        }

        // Add downstream callers (functions that call this)
        for (const caller of downstreamCallers) {
            if (caller.code && !includedCode.has(caller.code)) {
                parts.push(caller.code);
                includedCode.add(caller.code);
            }
        }

        // Neutralize each part before joining — every one is scanned-repository
        // source and may forge the separator.
        return parts.map(neutralizeBoundaries).join(FILE_BOUNDARY);
    }

    /**
     * Collect unique file paths from primary and all dependencies
     */
    _collectFilesIncluded(primaryFilePath, upstreamDependencies, downstreamCallers) {
        const files = new Set();
        files.add(primaryFilePath);

        // Separator is the FIRST colon (matches the dependency_resolver
        // contract); functionName may itself contain colons.
        for (const dep of upstreamDependencies) {
            if (dep.id) {
                const colonIndex = dep.id.indexOf(':');
                if (colonIndex > 0) {
                    files.add(dep.id.substring(0, colonIndex));
                }
            }
        }

        for (const caller of downstreamCallers) {
            if (caller.id) {
                const colonIndex = caller.id.indexOf(':');
                if (colonIndex > 0) {
                    files.add(caller.id.substring(0, colonIndex));
                }
            }
        }

        return Array.from(files);
    }

    /**
     * Create a single analysis unit
     */
    _createUnit(functionId, funcData, routeHandlerMap) {
        // Parse function ID: "relative/path/file.ts:functionName".
        // The separator is the FIRST colon: the functionName may itself contain
        // colons (e.g. an Express route id "src/r.ts:express(GET:/items/:id)").
        // This matches the dependency_resolver contract (`funcId.split(':')[0]`)
        // and recovers filePath/functionName correctly for multi-colon ids.
        const colonIndex = functionId.indexOf(':');
        const filePath = functionId.substring(0, colonIndex);
        const functionName = functionId.substring(colonIndex + 1);

        // Check if this is a route handler
        let routeData = null;
        let unitType = funcData.unitType || 'function';

        // Try to match against route info
        if (routeHandlerMap.has(functionId)) {
            routeData = routeHandlerMap.get(functionId);
            unitType = 'route_handler';
        } else if (routeHandlerMap.has(functionName)) {
            routeData = routeHandlerMap.get(functionName);
            unitType = 'route_handler';
        }

        // If the analyzer attached Express route metadata directly to the
        // function (anonymous arrow handler / middleware), surface it on the
        // unit's `route` field even when no external routes.json was given.
        if (!routeData && funcData.routeMetadata) {
            const meta = funcData.routeMetadata;
            routeData = {
                method: meta.http_method,
                path: meta.http_path,
                handler: funcData.name,
                middleware: meta.named_middleware || [],
            };
        }

        // Get upstream dependencies (functions this calls)
        const upstreamIds = this.resolver.getDependencies(functionId);
        const upstreamDependencies = [];

        for (const depId of upstreamIds) {
            const depFunc = this.resolver.functions[depId];
            if (depFunc) {
                upstreamDependencies.push({
                    id: depId,
                    name: depFunc.name,
                    code: depFunc.code,
                    unit_type: depFunc.unitType || 'function',
                    class_name: depFunc.className || null
                });
            }
        }

        // Get downstream callers (functions that call this)
        const callerIds = this.resolver.getCallers(functionId);
        const downstreamCallers = [];

        for (const callerId of callerIds) {
            const callerFunc = this.resolver.functions[callerId];
            if (callerFunc) {
                downstreamCallers.push({
                    id: callerId,
                    name: callerFunc.name,
                    code: callerFunc.code,
                    unit_type: callerFunc.unitType || 'function',
                    class_name: callerFunc.className || null
                });
            }
        }

        // Get direct calls from call graph
        const directCalls = this.resolver.callGraph[functionId] || [];
        const directCallers = this.resolver.reverseCallGraph[functionId] || [];

        // Assemble enhanced code with dependencies (VulnFounder standard format)
        const filesIncluded = this._collectFilesIncluded(filePath, upstreamDependencies, downstreamCallers);
        const hasDepsInlined = upstreamDependencies.length > 0 || downstreamCallers.length > 0;
        const assembledCode = this._assembleEnhancedCode(funcData, upstreamDependencies, downstreamCallers);

        const unit = {
            id: functionId,
            unit_type: unitType,
            code: {
                primary_code: assembledCode,
                primary_origin: {
                    file_path: filePath,
                    start_line: funcData.startLine || null,
                    end_line: funcData.endLine || null,
                    function_name: funcData.name,
                    class_name: funcData.className || null,
                    deps_inlined: hasDepsInlined,
                    files_included: filesIncluded,
                    original_length: funcData.code.length,
                    enhanced_length: assembledCode.length
                },
                dependencies: [],
                dependency_metadata: {
                    depth: this.maxDepth,
                    total_upstream: upstreamDependencies.length,
                    total_downstream: downstreamCallers.length,
                    direct_calls: directCalls.length,
                    direct_callers: directCallers.length
                }
            },
            route: routeData ? {
                method: routeData.method,
                path: routeData.path,
                handler: routeData.handler,
                middleware: routeData.middleware || []
            } : null,
            is_entry_point: funcData.isEntryPoint === true ? true : undefined,
            http_method: funcData.routeMetadata ? funcData.routeMetadata.http_method : undefined,
            http_path: funcData.routeMetadata ? funcData.routeMetadata.http_path : undefined,
            callback_index: funcData.routeMetadata ? funcData.routeMetadata.callback_index : undefined,
            ground_truth: {
                status: 'UNKNOWN',
                vulnerability_types: [],
                issues: [],
                annotation_source: null,
                annotation_key: null,
                notes: null
            },
            metadata: {
                is_exported: funcData.isExported || false,
                export_type: funcData.exportType || null,
                generator: 'unit_generator.js',
                direct_calls: directCalls,
                direct_callers: directCallers
            }
        };

        return unit;
    }
}

// CLI interface
if (require.main === module) {
    const args = process.argv.slice(2);

    if (args.length < 1) {
        console.error('Usage: node unit_generator.js <typescript_analyzer_output.json> [options]');
        console.error('');
        console.error('Options:');
        console.error('  --output <file>       Write results to file instead of stdout');
        console.error('  --depth <N>           Max dependency resolution depth (default: 3)');
        console.error('  --routes <routes.json> Route information from ast_parser.js');
        console.error('  --name <name>         Dataset name (default: derived from repo path)');
        console.error('');
        console.error('Examples:');
        console.error('  node unit_generator.js analyzer_output.json --output dataset.json --depth 2');
        process.exit(1);
    }

    const analyzerFile = args[0];
    let outputFile = null;
    let routesFile = null;
    let datasetName = null;
    let maxDepth = 3;

    // Parse arguments
    for (let i = 1; i < args.length; i++) {
        if (args[i] === '--output' && args[i + 1]) {
            outputFile = args[i + 1];
            i++;
        } else if (args[i] === '--depth' && args[i + 1]) {
            maxDepth = parseInt(args[i + 1], 10);
            i++;
        } else if (args[i] === '--routes' && args[i + 1]) {
            routesFile = args[i + 1];
            i++;
        } else if (args[i] === '--name' && args[i + 1]) {
            datasetName = args[i + 1];
            i++;
        }
    }

    try {
        // Load analyzer output
        if (!fs.existsSync(analyzerFile)) {
            console.error(`Analyzer output file not found: ${analyzerFile}`);
            process.exit(1);
        }
        const analyzerOutput = JSON.parse(fs.readFileSync(analyzerFile, 'utf-8'));

        // Load routes if provided
        let routeInfo = null;
        if (routesFile && fs.existsSync(routesFile)) {
            routeInfo = JSON.parse(fs.readFileSync(routesFile, 'utf-8'));
        }

        // Determine repo path from analyzer output or use current directory
        const repoPath = analyzerOutput.repoRoot || analyzerOutput.repository || process.cwd();

        // Generate units with dependency resolution
        console.error(`Processing ${Object.keys(analyzerOutput.functions || {}).length} functions...`);
        console.error(`Dependency resolution depth: ${maxDepth}`);

        const generator = new UnitGenerator(repoPath, datasetName, { maxDepth });
        const result = generator.generateUnits(analyzerOutput, routeInfo);

        // Merge with existing dataset if present
        let finalResult = result;
        if (outputFile && fs.existsSync(outputFile)) {
            try {
                const existing = JSON.parse(fs.readFileSync(outputFile, 'utf-8'));
                const existingUnits = existing.units || [];
                const existingIds = new Set(existingUnits.map(u => u.id));

                // Find new units not already in existing dataset
                const newUnits = result.units.filter(u => !existingIds.has(u.id));
                const duplicateCount = result.units.length - newUnits.length;

                console.error(`Existing dataset found:`);
                console.error(`  Existing units: ${existingUnits.length}`);
                console.error(`  New units to add: ${newUnits.length}`);
                console.error(`  Duplicates skipped: ${duplicateCount}`);
                if (duplicateCount > 0) {
                    console.error(`  Note: ${duplicateCount} existing units kept as-is (use 'vulnfounder parse --fresh' to regenerate all units)`);
                }

                // Append new units to existing
                finalResult = {
                    ...existing,
                    units: [...existingUnits, ...newUnits],
                    statistics: {
                        totalUnits: existingUnits.length + newUnits.length,
                        byType: Object.create(null),
                        callGraph: existing.statistics?.callGraph || {}
                    },
                    metadata: {
                        ...existing.metadata,
                        lastMerged: new Date().toISOString(),
                        lastAddedUnits: newUnits.length
                    }
                };

                // Recalculate byType statistics
                for (const unit of finalResult.units) {
                    const t = unit.unit_type || 'function';
                    finalResult.statistics.byType[t] = (finalResult.statistics.byType[t] || 0) + 1;
                }
            } catch (e) {
                console.error(`Could not parse existing file, will overwrite: ${e.message}`);
            }
        }

        const output = JSON.stringify(finalResult, null, 2);

        if (outputFile) {
            fs.writeFileSync(outputFile, output);
            console.error(`Dataset generated. Results written to: ${outputFile}`);
            console.error(`Total units: ${result.statistics.totalUnits}`);
            console.error(`By type:`, result.statistics.byType);
            console.error(`Call graph: ${result.statistics.callGraph.totalEdges} edges, avg degree: ${result.statistics.callGraph.avgOutDegree}`);
        } else {
            console.log(output);
        }

        process.exit(0);
    } catch (error) {
        console.error(`Error: ${error.message}`);
        process.exit(1);
    }
}

module.exports = { UnitGenerator, FILE_BOUNDARY, neutralizeBoundaries };
